[Tian Xiao] iP

snoidetx opened the pull request:
# Ultoi

> "I am nobody; my mind is everybody." - Snoide

Let `Ultoi` be your personal task manager. It's,
* text-based
* easy to learn
* ***UNBELIEVABLY FAST*** to use

All you need to do is,
1. download it from [here]().
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 🦦

And it is ***FREE***!

Features:
- [x] Managing to-dos, deadlines and events
- [ ] Reminders (coming soon...)

```
hello ultoi
```

Commit:
Add Gradle support

Commit:
Incremented Level-1.

Commit:
Incremented Level-2.

Commit:
Incremented Level-3 (Mark as Done).

Commit:
Increment Level-4.

Commit:
Incremented Level-5 (handle errors).

Commit:
Incremented Level-6 (Delete).

Commit:
Updated the comment for CheckInput.

Commit:
Updated Ultoi.java.

Commit:
Tried A-TextUITesting. Add a Command class to prepare for further OOP.

Commit:
Increment Level-7.

Commit:
Increment Level-8.

Commit:
Merge branch 'branch-level-8'

# Conflicts:
#	src/main/java/Ultoi.java

Commit:
no message

Commit:
Increment A-MoreOOP

Commit:
Increment A-Packages

Commit:
Add ListCommandTest.java

Commit:
Add ListCommandTest.java

Commit:
Conduct unit test

Commit:
Add Level-9 Find

Commit:
Add JavaDoc for all public classes and methods

Commit:
Update coding standard

Commit:
Merge branch 'Branch-Level-9' into Branch-A-CodingStandard

Commit:
Merge branch 'Branch-A-JavaDoc' into Branch-A-CodingStandard

# Conflicts:
#	src/main/java/ultoi/task/Task.java
#	src/main/java/ultoi/util/DateTime.java
#	src/main/java/ultoi/util/Parser.java
#	src/main/java/ultoi/util/TaskList.java

Commit:
Fix bugs and add wrongDateTimeException method

isabelteo commented on src/main/java/ultoi/util/DateTime.java:
```diff
@@ -0,0 +1,56 @@
+            throw UltoiException.wrongDateTimeFormatException();
+        }
+
+        if (this.time < 0 || time % 100 > 59 || time / 100 > 23) {
```
Maybe you could consider using constants for 59 and 23 to point out what the numbers mean?

isabelteo commented on src/main/java/ultoi/util/DateTime.java:
```diff
@@ -0,0 +1,56 @@
+            throw UltoiException.wrongDateTimeFormatException();
+        }
+
+        if (this.time < 0 || time % 100 > 59 || time / 100 > 23) {
```
Maybe you could consider using constants for 59 and 23 to point out what the numbers mean?

snoidetx replied:
That's a great suggestion! I have been using constants throughout my work and that is something I have looked :)

isabelteo commented on src/main/java/ultoi/util/DateTime.java:
```diff
@@ -0,0 +1,56 @@
+     */
+    @Override
+    public String toString() {
+        return this.date.toString() + " " + (this.time > 999 ? "" : "0")
```
maybe you could consider changing some of the ternary operators into boolean variables to improve readability?

isabelteo commented on src/main/java/ultoi/util/Parser.java:
```diff
@@ -0,0 +1,127 @@
+        try {
+            switch (tokens[0]) {
+            case "list":
+                return new ListCommand(input);
```
maybe you can consider adding `break;` to the end of each case or the comment `//Fallthrough`

isabelteo commented on src/main/java/ultoi/util/UltoiException.java:
```diff
@@ -0,0 +1,77 @@
+ * Represents the exceptions that could arise during the execution of  Ultoi bot.
+ */
+public class UltoiException extends Exception {
+    public static final String EXCEPTION_FACE = "<OnO>";
```
great use of prefixes here to make the constants easily readable!

snoidetx replied:
Thank you!

isabelteo reviewed (commented):
Overall, I think the code is very succinct and efficient as the methods are not deeply nested or overly complicated. I just commented on some small things that you could consider to improve readability, but great job :+1:

Commit:
Merge remote-tracking branch 'origin/add-gradle-support'

Commit:
Increment A-Gradle

lzan98 commented on src/main/java/ultoi/command/Command.java:
```diff
@@ -0,0 +1,24 @@
+ * @version 0.0.0
+ */
+public interface Command {
+    public void execute(UltoiUi ui, TaskList tasks, Storage storage) throws UltoiException;
```
I like how you used an interface for other commands to implement.

lzan98 commented on src/main/java/ultoi/command/MarkCommand.java:
```diff
@@ -0,0 +1,82 @@
+     * @throws UltoiException If the input cannot be identified.
+     */
+    public MarkCommand(String input) throws UltoiException {
+        this.isMark = input.startsWith("mark");
```
Clever way to combine mark and unmark commands into one class by making use of the boolean isMark.

Commit:
Increment A-CheckStyle

lzan98 commented on src/main/java/ultoi/util/TaskList.java:
```diff
@@ -0,0 +1,145 @@
+        return "Now you have " + size() + " task(s) in total.";
+    }
+
+    public TaskList findTasksWith(String keyword) {
```
Possibly missing a JavaDoc comment here!

lzan98 commented on src/main/java/ultoi/util/Parser.java:
```diff
@@ -0,0 +1,127 @@
+        String[] tokens = input.split(" ");
+
+        try {
+            switch (tokens[0]) {
+            case "list":
+                return new ListCommand(input);
+            case "todo":
+            case "deadline":
+            case "event":
+                return new AddCommand(input);
+            case "delete":
+                return new DeleteCommand(input);
+            case "mark":
+            case "unmark":
+                return new MarkCommand(input);
+            case "find":
+                return new FindCommand(input);
+            case "bye":
+                return new ByeCommand(input);
+            default:
+                throw UltoiException.noSuchCommandException();
+            }
```
Shows a very good understanding of switch statements and fall through.

lzan98 reviewed (commented):
Excellent OOP principles were displayed, code was neat and organised . Only left out one JavaDoc comment otherwise it is as good as it gets. Learnt a few tricks from your code. Well done.

Commit:
Fix bugs related to load and save to hard disk

Commit:
Add basic GUI features

Commit:
Optimise GUI and fix bugs

Commit:
Implement Gradle

Commit:
Setup checkstyle

Commit:
Add GUI

Commit:
Merge branch 'Branch-A-CheckStyle'

Commit:
Merge branch 'Branch-Level-10'

Commit:
Change jar file name

Commit:
Add Java assertions

Currently, there are some assumptions made in the logic.

They may cause unintended consequences and regressions in the actual run.

Let's add several Java assertions to help us root the potential problems.

Using assertions is preferable because it narrows down the general error to a specific part of the code.

Commit:
Improve code quality

Commit:
Add Java streams

Some parts of the current code are implemented with for-loops.

It may decrease the readability and sometimes contain nested loops.

Let's change some of them to streams instead.

Streams process a sequence of data using lambda functions, reduce nested loops and improve readability.

snoidetx committed:
Merge pull request #2 from snoidetx/branch-A-CodeQuality

Improve code quality

snoidetx committed:
Revert "Improve code quality"

snoidetx committed:
Merge pull request #4 from snoidetx/revert-2-branch-A-CodeQuality

Revert "Improve code quality"

Commit:
Merge branch 'master' of https://github.com/snoidetx/ip into branch-A-Streams

Commit:
Merge commit 'e40e33fbdfc658eeaa881c1577e51311eca7d834'

Commit:
Merge branch 'branch-A-Streams'

snoidetx committed:
Create main.yml

Commit:
no message

Commit:
Update SortCommand.java

snoidetx committed:
Merge pull request #5 from snoidetx/branch-NewMaster

Branch new master

Commit:
Update version number

Commit:
Merge branch 'branch-RestoreCodeQuality1'

# Conflicts:
#	src/main/java/ultoi/command/AddCommand.java

Commit:
Restore changes in coding style

Commit:
Increment C-Sort

Commit:
Improve GUI

Commit:
Add error box and upload product representative picture

Commit:
Update README.md

Commit:
Fix bugs

snoidetx committed:
Rename main.yml to gradle.yml

snoidetx committed:
Update README.md

snoidetx committed:
Rename README.md to UserGuide.md

snoidetx committed:
Update README.md
